java/vector/src/main/java/org/apache/arrow/vector/compare/RangeEqualsVisitor.java

```diff
@@ -49,38 +56,66 @@ public RangeEqualsVisitor(ValueVector right, int leftStart, int rightStart, int
     this.rightStart = rightStart;
     this.right = right;
     this.length = length;
+    Preconditions.checkArgument(length >= 0, "length must be non negative");
+  }
+
+  /**
+   * Reset start indices and length for reuse purpose.
+   */
+  public void reset(int leftStart, int rightStart, int length) {
+    this.leftStart = leftStart;
+    this.rightStart = rightStart;
+    this.length = length;
+  }
+
+  private boolean validateTypesAndIndices(ValueVector left) {
+
+    if (needCheckType && !compareValueVector(left, right)) {
```

Comment:
this assumes that `compareValueVector` only checks the type. This should be called regardless of the `needCheckType` flag, and then in `compareValueVector` you can look at the flag before comparing types

Reply:
Right, fixed.